rest_framework/serializers.py

```diff
@@ -851,7 +851,97 @@ def raise_errors_on_nested_writes(method_name, serializer, validated_data):
     )
 
 
-class ModelSerializer(Serializer):
+class ModelSerializerMetaclass(SerializerMetaclass):
+    """
+    This metaclass sets a dictionary named `base_fields` on the class.
+
+    `base_fields` include the declared fields (see SerializerMetaclass),
+    and fields generated for the model. A deepcopy of `base_fields` is
+    used for each instance of the ModelSerializer.
+    """
+
+    @classmethod
+    def _get_fields(cls, new_class):
+        # XXX
```

Comment:
I ran into some problem here. Consider we have `ModelSerializer`, `HyperlinkedModelSerializer`, and suppose we have a `CustomizedModelSerializer` (a custom variant of `ModelSerialzier`, like `HyperlinkedModelSerializer`) and then an actual `MyModelSerializer` (a model serializer for some `MyModel`).

All are subclasses of `ModelSerializer`, so the metaclass code to get the fields runs for all of them. However, we only want it to run for `MyModelSerializer` and not the others. The easiest way to do it is check whether `Meta` is defined, but then we lose the `Class {serializer_class} missing "Meta" attribute` check which is quite useful for new users.

Now that I think about it, we can set `base_fields = None` if `Meta` is not defined, and then raise the error in `get_fields()`. This will be similar to the current behavior - error on first instantiation. WDYT?

Reply:
I went ahead and made this change.